src/app/data-planes/views/DataPlaneListView.spec.ts

```diff
@@ -2,10 +2,24 @@ import { describe, expect, test } from '@jest/globals'
 import { flushPromises, mount } from '@vue/test-utils'
 
 import DataPlaneListView from './DataPlaneListView.vue'
-import { router } from '@/../jest/jest-setup-after-env'
-
-async function renderComponent() {
-  await router.push({ name: 'data-plane-list-view', params: { mesh: 'default' } })
+import { router, useMock } from '@/../jest/jest-setup-after-env'
+import response from '@/api/mock-data/meshes/default/dataplanes+insights.json'
+
+const mock = useMock('/meshes/default/dataplanes+insights', response)
+
+async function renderComponent(props?: Record<string, unknown>) {
+  await router.push(
+    {
+      ...{
+        name: 'data-plane-list-view',
+        params: { mesh: 'default' },
+        meta: {
+          type: 'standard',
+        },
+      },
+      ...props || {},
+    },
+  )
```

Comment:
**Suggestion**:

* Avoid calling this `props` to not confuse the initial route with the quasi-default `props` parameter passed to other `renderComponent` functions.
* Use vue-router’s `RouteLocationNamedRaw` type to pass complete usable route objects

```suggestion
const DEFAULT_ROUTE: RouteLocationNamedRaw = {
  name: 'data-plane-list-view',
  params: {
    mesh: 'default',
  },
  meta: {
    type: 'standard',
  },
}

async function renderComponent(initialRoute: RouteLocationNamedRaw = DEFAULT_ROUTE) {
  await router.push(initialRoute)
```

Reply:
I think the reason why passing `meta` to a `router.push` doesn’t work with the types available is that `meta` is part of what _defines_ a route but not part of the things _identify_ a route **in navigation**. In other words: `router.push` accepts things that are represented in the URL and meta isn’t taken into account when vue-router identifies a matching route. I’m pretty sure you could call `router.push` with an arbitrary value for `meta` (assuming one were to disable type checking for that call) and it wouldn’t break that call as long as the rest of the route location identifies an existing route.

What you’re trying in this test is "navigate to the data plane list view but act as if its meta data field `type` is `gateway`". Instead, you should just ask the test to navigate to the `gateway-list-view`.

**Setup**:

```suggestion
const DEFAULT_ROUTE: RouteLocationNamedRaw = {
  name: 'data-plane-list-view',
  params: {
    mesh: 'default',
  },
}
async function renderComponent(initialRoute: RouteLocationNamedRaw = DEFAULT_ROUTE) {
  await router.push(initialRoute)
```

**Usage**:

```ts
const wrapper = await renderComponent({
  name: 'gateway-list-view',
  params: {
    mesh: 'default',
  },
})
```

---

> Not quite sure what you meant by this? Are you referring to the name of the local argument for the function? If so happy to change the name if it's just that, but it's a local variable so I don't think it matters that much, how about `partialRouteLocation`?

My point is merely one of naming. Props has a very fixed and defined meaning in a Vue application and the parameter expected here is _not_ a prop but a route location.

Reply:
Ah I getcha, I think. Is what you are saying is I just need to change the `name` to either 'dataplane-list-view' or 'gateway-list-view' here? Then whatever the `meta` is in the routing configuration in router.ts will then be applied?

I'll give that a go.

As an aside/observation, if I've understood you correctly, this definitely feels like we are testing the entire routing/app rather than just the isolated component. All in all I think I'd rather just use component props, I think I tried that before trying out meta, might give that another go instead also 🤔

Reply:
> Ah I getcha, I think. Is what you are saying is I just need to change the name to either 'dataplane-list-view' or 'gateway-list-view' here? Then whatever the meta is in the routing configuration in router.ts will then be applied? i.e. I shouldn't be setting meta at all.

Yes, that’s correct. `router.push` will resolve the route location provided to it against the route definitions the router instances has registered. It will try to match by name and params to find the correct definition and use that as the route meaning that the meta information will be set correctly.

> As an aside/observation, if I've understood you correctly, this definitely feels like we are testing the entire routing/app rather than just the isolated component.

I don’t think that’s the case. We have some components that just need access to route-specific information and for that, it’s important to supply this in the tests.

In this test, we navigate to the data-plane-list-view in the router because that’s accurate, but we only do that because of various access to `route` information like the mesh parameter or now the meta `type` property. This doesn’t test whether the navigation takes places. Think of it more like what you’d do with mocking. You can achieve similar results by mounting the component with global mocks:

```js
mount(Component, {
  global: {
    mocks: {
      // Works for accessing `this.$router`, won’t work for accessing `router` that comes from `useRouter`.
      $router: {
        params: { mesh: 'default' },
      },
    },
  },
})

Reply:
Given that we don’t have an end-to-end test environment (yet), I care less about the philosophical purity of the test approach. As long as the test works reasonably well and does relevant assertions, it’s good for me.

Technically speaking, these tests can’t be end-to-end tests because they’re run in a Node.js environment with DOM simulation and all that.

Reply:
Right yeah, so these should aim to be isolated component rather than e2e

Reply:
Yes, but calling `router.push` here doesn’t make them end-to-end tests or less isolated.

Reply:
In my mind it's depending on router functionality, which makes the component test less isolated. But whichever, I don't think it matters too much, but I'm gonna have a look a removing the router either via mocking or look again at trying to pass props straight into the component, even if its just for the learning experience.